text/0000-rand-crate-redesign.md

```diff
@@ -0,0 +1,1063 @@
+purpose.
+
+The strawman revision adds several new distributions, namely `Uniform`,
+`Uniform01`, and `Default` (see below). These distributions allow creation of
+random values for arbitrary types, replacing the need for a `Rand` trait. The
+[`Rand`] and [`SimpleRand`] traits in the strawman revision are simply
+wrappers around distributions, specifically `Default`.
+
+Additionally, the strawman revision adds a trait called [`Sample`]. This is
+simply an extension to [`Rng`], adding some convenience functions to access
+functionality from the `Default` and `Range` distributions, as well as
+iterators.
+
+So the strawman revision currently has *three* convenience wrappers around
+distributions: [`Rand`], [`SimpleRand`] and [`Sample`]. My feeling is that the
+most convenient of these is [`Sample`] and the other two serve no real purpose
+(note that implementing random value creation for user-defined types is
+now handled via `impl Distribution<T> for Default`).
+
+#### Examples
+
+Some [`Rand`] examples:
+
+```rust
+use rand::distributions::{Rand, Default, Range};
+let mut rng = rand::thread_rng();
+
+// Type annotation needed; two options:
+let byte: u8 = Rand::rand(&mut rng, Default);
+let byte = u8::rand(&mut rng, Default);
+
+// For ranges, the generated type is the same as the parameter type:
+let ranged = Rand::rand(&mut rng, Range::new(-99, 100));
+```
+
+Some [`SimpleRand`] examples:
+
+```rust
+use rand::distributions::{SimpleRand, Distribution, Range};
+let mut rng = rand::thread_rng();
+
+// Again, type annotation is needed; two options:
+let byte: u8 = SimpleRand::simple_rand(&mut rng);
+let byte = u8::simple_rand(&mut rng);
+
+// SimpleRand does not support other distributions, so we have to use the
+// distribution directly:
+let ranged = Range::new(-99, 100).sample(&mut rng);
+```
+
+Some [`Sample`] examples:
+
+```rust
+use rand::distributions::{Sample, Rand, Default, Range};
+let mut rng = rand::thread_rng();
+
+// Type annotation needed:
+let byte: u8 = rng.gen();
+
+// For ranges, the generated type is the same as the parameter type:
+let ranged = rng.gen_range(-99, 100);
+```
+
+Equivalent code without using any of the wrappers:
+
+```
+use rand::distributions::{Distribution, Default, Range};
+let mut rng = rand::thread_rng();
+
+let byte: u8 = Default.sample(&mut rng);
+
+let ranged = Range::new(-99, 100).sample(&mut rng);
+```
+
+#### Pass by copy?
+
+Currently [`Rand::rand`] and [`Sample::sample`] take the distribution parameter
+by value. This is the best option for zero-size distribution types like
+[`Default`] and [`Open01`], since it allows call syntax like
+`Rand::rand(&mut rng, Default)` (second parameter
+does not need to be referenced).
+
+Most distribution types are fairly small, e.g. `Range` is two or three values
+of the parameterised type, so for the most part pass-by-value is reasonable,
+although for example `Gamma` is 40 bytes. Can the compiler optimise this?
```

Comment:
Currently the `Gamma` distribution is very slow, 3~4 times worse than the other distributions. I can't image pass-by-value to be a large contributing factor here.

Maybe the implementation can be replaced with some better algorithm. Like the ziggurat method used in the other distributions. That should take care of the large parameter also.

Anyway, I wouldn't worry about pass-by-value here.